### ISSUE TITLE
possible Instant clock skew detected

### ISSUE THREAD

**Issue by wendajiang** [tokio-rs/console#299]
I use console_subscriber  for tracing my app, and terminal sometimes output the `possible Instant clock skew detected`, and I find #286. So I confuse about why after use Instant instead of SystemTime, this still happend? And what is meaning?

**Comment by hawkw** [comment]
This message indicates that polling a task appears to have ended before it started: https://github.com/tokio-rs/console/blob/abc083000cb6de51e37d5037283e97ed0e27249e/console-subscriber/src/stats.rs#L493-L503

`Instant::now` type is _supposed_ to be monotonic, but sometimes there are platform-dependent issues in the implementation of `Instant` for particular operating systems or hardware architectures. What platform is the instrumented application running on?

Alternatively, there might be a bug in the `console-subscriber` crate. In order to determine what's going on here, it would be very helpful to know some details about the operating system and hardware architecture the instrumented application (not the `tokio_console` CLI) is running on, as well as the frequency of the warning message.

**Comment by wendajiang** [comment]
Platform is macOS, arch is arm64. And in tui-rs application, once app is opening, about ten warning per minute.

**Comment by hawkw** [comment]
Interesting, thanks. Can you possibly also paste the complete text of one or two of the warnings, including the start and end timestamps? It would be useful to see how much skew was observed.

**Comment by Rexagon** [comment]
Faced a similar error. These messages appear consistently every few minutes.

```
16:08:08 ...
possible Instant clock skew detected: a poll's end timestamp was before its start timestamp
start = Instant { tv_sec: 4845, tv_nsec: 962726114 }
  end = Instant { tv_sec: 4845, tv_nsec: 962723900 }

...

16:08:19 ...
possible Instant clock skew detected: a poll's end timestamp was before its start timestamp
start = Instant { tv_sec: 4857, tv_nsec: 520769429 }
  end = Instant { tv_sec: 4857, tv_nsec: 520769068 }

...

16:08:19 ...
possible Instant clock skew detected: a poll's end timestamp was before its start timestamp
start = Instant { tv_sec: 4857, tv_nsec: 727735375 }
  end = Instant { tv_sec: 4857, tv_nsec: 727733632 }

...

16:08:23 ...
possible Instant clock skew detected: a poll's end timestamp was before its start timestamp
start = Instant { tv_sec: 4861, tv_nsec: 607221176 }
  end = Instant { tv_sec: 4861, tv_nsec: 607220114 }
```

Env info:
```
Linux 5.17.1-3-MANJARO
AMD Ryzen 9 5900HS

rustc 1.60.0 (7737e0b5c 2022-04-04)

tokio 1.18.0 (full, tracing, parking_lot)
tracing 0.1.34
console-subscriber 0.1.5 (parking_lot)
```

**Comment by zbuc** [comment]
We are seeing this as well:

```
possible Instant clock skew detected: a poll's end timestamp was before its start timestamp
start = Instant { tv_sec: 4729132, tv_nsec: 88408831 }
  end = Instant { tv_sec: 4729132, tv_nsec: 88406315 }
possible Instant clock skew detected: a poll's end timestamp was before its start timestamp
start = Instant { tv_sec: 4729243, tv_nsec: 938458369 }
  end = Instant { tv_sec: 4729243, tv_nsec: 938455058 }
possible Instant clock skew detected: a poll's end timestamp was before its start timestamp
start = Instant { tv_sec: 4729793, tv_nsec: 955538207 }
  end = Instant { tv_sec: 4729793, tv_nsec: 955531205 }
possible Instant clock skew detected: a poll's end timestamp was before its start timestamp
start = Instant { tv_sec: 4730224, tv_nsec: 59384090 }
  end = Instant { tv_sec: 4730224, tv_nsec: 59377093 }
```

Env:
```
Linux dd0426221be7 5.4.0-100-generic #113-Ubuntu SMP Thu Feb 3 18:43:29 UTC 2022 x86_64 GNU/Linux
rustc 1.60.0 (7737e0b5c 2022-04-04)

tokio 1.18.1 (full)
tracing 0.1.34
console-subscriber 0.1.5
```

**Comment by Rexagon** [comment]
I think there is no guarantee that two `Instant`s created sequentially on different cores will also be sequential.

https://stackoverflow.com/questions/46893072/clock-gettimeclock-monotonic-monotonicity-across-cores-threads

**Comment by hawkw** [comment]
@Rexagon hmm...in this case, a poll's start and end timestamps should be recorded on the same core. The poll start timestamp corresponds to the beginning of a call to the `poll` function, and the poll end timestamp corresponds to the end of a call to that function — the task cannot cross threads while it's being polled...

**Comment by Rexagon** [comment]
Yes, `span::Entered` is not `Send`, so this skew is strange. But maybe one of the calls of the `PollStats::start_poll` or its contents are missing somewhere under strange conditions?

**Comment by hawkw** [comment]
@Rexagon yeah, if I had to guess, i would suspect it's something more like that. Hmmm..

**Comment by izderadicka** [comment]
Same problem here - Linux Ubuntu 20.04 on Intel.
Having these printout quite often - several times per second - when my program (TCP proxy) is under load.

```
possible Instant clock skew detected: a poll's end timestamp was before its start timestamp
start = Instant { tv_sec: 255185, tv_nsec: 521611026 }
  end = Instant { tv_sec: 255185, tv_nsec: 521610916 }
possible Instant clock skew detected: a poll's end timestamp was before its start timestamp
start = Instant { tv_sec: 255185, tv_nsec: 720152934 }
  end = Instant { tv_sec: 255185, tv_nsec: 720127010 }
possible Instant clock skew detected: a poll's end timestamp was before its start timestamp
start = Instant { tv_sec: 255186, tv_nsec: 177118060 }
  end = Instant { tv_sec: 255186, tv_nsec: 177056943 }
```
Difference can be several micros.

console-subscriber = "0.1.8"

If you are interested it can be easily replicated.

**Comment by thomasj02** [comment]
Also getting this frequently on 0.1.8, with Ubuntu 22.04 on Intel

**Comment by ozgunozerk** [comment]
Also getting this frequently on macOS

**Comment by jcreekmore** [comment]
I see this as well and here is what I think is happening (relevant bits: `tokio` v1.47.1, `console-subscriber` v0.4.1, and `sqlx-sqlite` v0.8.6).

1. I ran my application with RUST_BACKTRACE=1 and modifying `console-subscriber` to panic after printing the clock skew message. The backtrace showed "thread 'sqlx-sqlite-worker-27' panicked". I'm running my application on a rt-multi-thread `tokio` runtime and I have 32 cores on this box. 
2. We have a tokio task where we have a span created and then we instrument an async function. Inside of that async function, we do several `sqlx` requests.
3. A command channel for the worker thread for `sqlx-sqlite` in the `establish` method (seen here https://github.com/launchbadge/sqlx/blob/3abb1863244685eabc274eae665fe1c6a300a36e/sqlx-sqlite/src/connection/worker.rs#L106-L348) and the command channel transmit side is stored in the connection that is then passed out of the worker thread on line 133 and returned on line 347. With the command channel set up, the worker thread then starts iterating across commands on line 147 where it gets a command and a span and then it enters the span on line 148. So, we are entering and exiting the span inside of the worker thread. But where does that span come from (other than the command channel?)
4. If we take the `execute` method on the connection (https://github.com/launchbadge/sqlx/blob/3abb1863244685eabc274eae665fe1c6a300a36e/sqlx-sqlite/src/connection/worker.rs#L360-L385), that method sends on the command channel the command and `Span::current()`. So, in our async context, we are getting our current span and sending a copy of it to the worker thread for this connection so that any trace messages from the database connection will get attached to the span.
5. My suspicion is that I have that span from the instrumentation mentioned in 2 that was entered when the async function was last woken, then we pass the span across the `flume::bounded` channel and then we race on the enter for the span (because we are probably on two separate physical cores, one for the async runtime portion and one for the worker thread).
6. So, if that is the case, then from my tokio task, the `start_poll` method was called, we increment `current_polls` to 1, lock the timestamps, and update `last_poll_started`. Then we still (if we are racing in 5) have not exited the span from the tokio task yet.
7. We enter the span in the worker thread, the `start_poll` method was called, but `current_polls` was previously 1, so we return without updating `last_poll_started`. 
8. THEN, the exit from the tokio task happens and the `end_poll` method is called. But, we have incremented `current_polls` to 2 at this point, so we return early from `end_poll`. 
9. Finally, we finish working in the worker thread for `sqlx`. The `_guard` goes out of scope and we again call `end_poll`. This time, we had decremented `current_polls` so we will treat this as "ending the last current poll" and we process the rest of the method. But, if we cannot guarantee monotonicity between cores (https://github.com/tokio-rs/console/issues/299#issuecomment-1117476308), then there is a very good chance that we will see the clock skew message. 

All of that said, I do not know what the right answer is here. I am fairly certain that this is the order of events that is happening and causing this message to be printed (at least in my application), but I am not certain one way or the other where the _fault_ actually lies. Since the `tracing::Span` is `Send`, it can freely be sent between threads.

**Comment by jcreekmore** [comment]
After digging around a bit more in the `console-subscriber` code, I saw that the `current_spans` was being tracked using a `ThreadLocal` and my original thinking yesterday was that these stats needed to be tracked local to the thread as well, so I made a basic change where I wrapped the `PollStats` in the `TaskStats` and `AsyncOpStats` with a `ThreadLocal` (I _also_ had to cache the `poll_duration_max` and `scheduled_duration_max` in the `TaskStats` so that I could construct the `PollStats` for the `TaskStats` in a thread local fashion). This change wasn't production ready because I kept the atomics and the mutex around and, since it is thread local, those could be changed so that the `PollStats` take a `&mut self` and wrap the whole thing in a `ThreadLocal<RefCell<PollStats>>` instead (again, similar to the `current_spans`).

That _seems_ to eliminate the `possible Instant clock skew detected: a poll's end timestamp was before its start timestamp` messages that I was seeing fairly regularly, at least in my limited testing so far.

Would this change (complete with moving it to `ThreadLocal<RefCell<PollStats>>`) be something you are interested in? Or am I completely off-base in my approach?
